### PR TITLE
Add val:*_pipeline,overrides,value,type_error/validation_error

### PR DESCRIPTION
**Pull request by shrekshao**
Issue: #2060

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.) (Pass with https://dawn-review.googlesource.com/c/dawn/+/112920 which is about to land; f16 is not fully implemented and is skipped;)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [x] Tests are properly located in the test tree.
- [x] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [x] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [x] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/api/validation/compute_pipeline.spec.ts [hunk 1]
```diff
@@ -368,6 +369,97 @@ Tests calling createComputePipeline(Async) validation for uninitialized overrida
...
+        { constants: { cu: kValue.u32.min }, _success: true },
+        { constants: { cu: kValue.u32.min - 1 }, _success: false },
+        { constants: { cu: kValue.u32.max }, _success: true },
+        { constants: { cu: kValue.u32.max + 1 }, _success: false },
+        { constants: { ci: kValue.i32.negative.min }, _success: true },
+        { constants: { ci: kValue.i32.negative.min - 1 }, _success: false },
+        { constants: { ci: kValue.i32.positive.max }, _success: true },
+        { constants: { ci: kValue.i32.positive.max + 1 }, _success: false },
+        { constants: { cf: kValue.f32.negative.min }, _success: true },
+        { constants: { cf: -Number.MAX_VALUE }, _success: false },
```
Comment (kainino0x) [inline src/webgpu/api/validation/compute_pipeline.spec.ts:428]: We can tighten all of the MAX_VALUE-to-f32 cases. `2**128` should be unambiguously invalid, though we can tighten it even more with `2**128 - 2**103`. That's the midpoint between FLT_MAX and the [special value](https://webidl.spec.whatwg.org/#es-float) of `2**128` which WebIDL says is supposed to be the first invalid value. And I think it should work in our implementation.

Aside: You can write this simply as `2**128 - 2**103` in JS. I verified with BigInt math:
```js
(2n**128n - 2n**104n + 2n**128n) / 2n === BigInt(2**128 - 2**103)
```

Of course we would like to test the next smaller float value (which is supposed to be valid), which is `2**128 - 2**105`, but this will fail in Chrome since we're not sure it's safe with ubsan. Given it's not very important, I say we leave it out for now and add a TODO to test it.

File: src/webgpu/api/validation/compute_pipeline.spec.ts [hunk 2]
```diff
@@ -368,6 +369,97 @@ Tests calling createComputePipeline(Async) validation for uninitialized overrida
...
+        { constants: { cf: kValue.f32.positive.max }, _success: true },
+        { constants: { cf: Number.MAX_VALUE }, _success: false },
+        // Conversion to boolean can't fail
+        { constants: { cb: Number.MAX_VALUE }, _success: true },
```
Comment (kainino0x) [inline src/webgpu/api/validation/compute_pipeline.spec.ts:432]: (can keep this MAX_VALUE)

File: src/webgpu/api/validation/compute_pipeline.spec.ts [hunk 3]
```diff
@@ -368,6 +369,97 @@ Tests calling createComputePipeline(Async) validation for uninitialized overrida
...
+        { constants: { cu: kValue.u32.max }, _success: true },
+        { constants: { cu: kValue.u32.max + 1 }, _success: false },
+        { constants: { ci: kValue.i32.negative.min }, _success: true },
+        { constants: { ci: kValue.i32.negative.min - 1 }, _success: false },
+        { constants: { ci: kValue.i32.positive.max }, _success: true },
+        { constants: { ci: kValue.i32.positive.max + 1 }, _success: false },
+        { constants: { cf: kValue.f32.negative.min }, _success: true },
+        { constants: { cf: -Number.MAX_VALUE }, _success: false },
+        { constants: { cf: kValue.f32.positive.max }, _success: true },
+        { constants: { cf: Number.MAX_VALUE }, _success: false },
```
Comment (kainino0x) [inline src/webgpu/api/validation/compute_pipeline.spec.ts:430]: Chrome doesn't support it yet but maybe we can add f16 tests now? They have the same rules as f32.

**Review by kainino0x — COMMENTED** [review]
(ditto comments in the vertex and fragment tests too)

File: src/webgpu/api/validation/compute_pipeline.spec.ts (hunk 1 above)
Comment (kainino0x) [reply src/webgpu/api/validation/compute_pipeline.spec.ts:428]: Sorry that was wrong. We want to test the next smaller `Number` (`double`) which is `2**128 - 2**104 - 2**75` not `2**128 - 2**105`. Editing

Comment (kainino0x) [reply src/webgpu/api/validation/compute_pipeline.spec.ts:428]:
```js
const kLastF64CastableToF32     = 340282356779733623858607532500980858880;
const kFirstF64NotCastableToF32 = 340282356779733661637539395458142568448;
```

File: src/webgpu/api/validation/compute_pipeline.spec.ts [hunk 4]
```diff
@@ -368,6 +369,97 @@ Tests calling createComputePipeline(Async) validation for uninitialized overrida
...
+  )
+  .params(u =>
+    u //
+      .combine('isAsync', [true, false])
+      .combineWithParams([
+        { constants: { cf: 1 }, _success: true }, // control
+        { constants: { cf: NaN }, _success: false },
+        { constants: { cf: Number.POSITIVE_INFINITY }, _success: false },
+        { constants: { cf: Number.NEGATIVE_INFINITY }, _success: false },
+      ] as { constants: Record<string, GPUPipelineConstantValue>; _success: boolean }[])
```
[the remark this answers is not in the document]
Comment (shrekshao) [reply src/webgpu/api/validation/compute_pipeline.spec.ts:386]: Yeah turns out "as const" works for this case since all entries are "cf"

[the remark this answers is not in the document]
Comment (kainino0x) [reply src/webgpu/api/validation/compute_pipeline.spec.ts:386]: Ahh yeah makes sense `as const` wouldn't work for the others.

File: src/webgpu/api/validation/compute_pipeline.spec.ts (hunk 1 above)
Comment (shrekshao) [reply src/webgpu/api/validation/compute_pipeline.spec.ts:428]: Oh I see. It can be get from removing one fraction bit
`340282356779733661637539395458142568448n - 2n**(0x47en - 1023n) / 2n**52n === 340282356779733623858607532500980858880n`

Comment (kainino0x) [reply src/webgpu/api/validation/compute_pipeline.spec.ts:428]: Yeah. I did this the bit-manipulation way:
```
> new BigUint64Array(new Float64Array([2**128 - 2**103]).buffer)[0]
5183643170835005440n
> BigInt(new Float64Array(new BigInt64Array([5183643170835005440n]).buffer)[0])
340282356779733661637539395458142568448n
```

File: src/webgpu/util/constants.ts [hunk 5]
```diff
@@ -372,11 +400,19 @@ export const kValue = {
       min: hexToF16(kBit.f16.positive.min),
       max: hexToF16(kBit.f16.positive.max),
       zero: hexToF16(kBit.f16.positive.zero),
+      first_f64_not_castable: hexToF16(kBit.f16.positive.max) / 2 + 2 ** 127, // mid point of 2**128 and largest f16
```
[the remark this answers is not in the document]
Comment (shrekshao) [reply src/webgpu/util/constants.ts]: Fix f16 boundary values; Add within/without f32 range cases to f16 tests

File: src/webgpu/api/validation/compute_pipeline.spec.ts [hunk 6]
```diff
@@ -368,6 +369,147 @@ Tests calling createComputePipeline(Async) validation for uninitialized overrida
...
+
+    t.doCreateComputePipelineTest(isAsync, _success, descriptor);
+  });
+
+g.test('overrides,value,validation_error,f16')
+  .desc(
+    `
+Tests calling createComputePipeline(Async) validation for unrepresentable f16 constant values in compute stage.
+
+TODO(#2060): test with last_f64_castable.
```
Comment (kainino0x) [inline src/webgpu/api/validation/compute_pipeline.spec.ts:470]:
```suggestion
TODO(#2060): Tighten the cases around the valid/invalid boundary once we have WGSL spec
clarity on whether values like f16.positive.last_f64_castable would be valid. See issue.
```

File: src/webgpu/api/validation/render_pipeline/overrides.spec.ts [hunk 7]
```diff
@@ -182,3 +183,317 @@ Tests calling createComputePipeline(Async) validation for uninitialized overrida
...
+
+    t.doCreateRenderPipelineTest(isAsync, _success, descriptor);
+  });
+
+g.test('value,validation_error,f16,vertex')
+  .desc(
+    `
+Tests calling createRenderPipeline(Async) validation for unrepresentable f16 constant values in vertex stage.
+
+TODO(#2060): test with last_f64_castable.
```
Comment (kainino0x) [inline src/webgpu/api/validation/render_pipeline/overrides.spec.ts:393]:
```suggestion
TODO(#2060): Tighten the cases around the valid/invalid boundary once we have WGSL spec
clarity on whether values like f16.positive.last_f64_castable would be valid. See issue.
```

File: src/webgpu/api/validation/render_pipeline/overrides.spec.ts [hunk 8]
```diff
@@ -182,3 +183,317 @@ Tests calling createComputePipeline(Async) validation for uninitialized overrida
...
+      },
+    });
+  });
+
+g.test('value,validation_error,f16,fragment')
+  .desc(
+    `
+Tests calling createRenderPipeline(Async) validation for unrepresentable f16 constant values in fragment stage.
+
+TODO(#2060): test with last_f64_castable.
```
Comment (kainino0x) [inline src/webgpu/api/validation/render_pipeline/overrides.spec.ts:449]:
```suggestion
TODO(#2060): Tighten the cases around the valid/invalid boundary once we have WGSL spec
clarity on whether values like f16.positive.last_f64_castable would be valid. See issue.
```

**Review by kainino0x — APPROVED** [review]
